From: Amit Shah <amit.shah@redhat.com>
Date: Thu, 18 Feb 2010 10:27:37 -0500
Subject: [kvm] virtio: Initialize vq->data entries to NULL
Message-id: <5bcf9013b8663e91718f9d606eb83e76047b2088.1266488420.git.amit.shah@redhat.com>
Patchwork-id: 23338
O-Subject: [PATCH 05/10] virtio: Initialize vq->data entries to NULL
Bugzilla: 566391
RH-Acked-by: Marcelo Tosatti <mtosatti@redhat.com>
RH-Acked-by: Juan Quintela <quintela@redhat.com>

vq operations depend on vq->data[i] being NULL to figure out if the vq
entry is in use.

We have to initialize them to NULL to ensure we don't work with junk
data and trigger false BUG_ONs.

Signed-off-by: Amit Shah <amit.shah@redhat.com>
Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/drivers/virtio/virtio_ring.c b/drivers/virtio/virtio_ring.c
index 71929ee..9bcfe95 100644
--- a/drivers/virtio/virtio_ring.c
+++ b/drivers/virtio/virtio_ring.c
@@ -431,8 +431,11 @@ struct virtqueue *vring_new_virtqueue(unsigned int num,
 	/* Put everything in free lists. */
 	vq->num_free = num;
 	vq->free_head = 0;
-	for (i = 0; i < num-1; i++)
+	for (i = 0; i < num-1; i++) {
 		vq->vring.desc[i].next = i+1;
+		vq->data[i] = NULL;
+	}
+	vq->data[i] = NULL;
 
 	return &vq->vq;
 }
